tiles/src/main/java/com/protomaps/basemap/layers/Roads.java

```diff
@@ -35,123 +42,163 @@ public void processFeature(SourceFeature sourceFeature, FeatureCollector feature
-          .setAttrWithMinzoom("ref_length", shield_text_length, 14)
-          .setAttrWithMinzoom("network", network_val, 14);
-        OsmNames.setOsmNames(feat, sourceFeature, 14);
+        kind = "other";
+        kind_detail = sf.getString("service");
+        min_zoom = 14;
+        min_zoom_shield_text = 14;
+        min_zoom_names = 14;
       }
 
-      if (sourceFeature.hasTag("highway", "motorway_link", "trunk_link", "primary_link", "secondary_link",
+      var feat = features.line("roads")
+        .setId(FeatureId.create(sf))
+        // Core Tilezen schema properties
+        .setAttr("pmap:kind", kind)
+        // To power better client label collisions
+        .setAttr("pmap:min_zoom", min_zoom + 1)
+        .setAttrWithMinzoom("ref", shield_text, min_zoom_shield_text)
+        .setAttrWithMinzoom("shield_text_length", shield_text_length, min_zoom_shield_text)
+        .setAttrWithMinzoom("network", network_val, min_zoom_shield_text)
+        // Core OSM tags for different kinds of places
+        .setAttrWithMinzoom("layer", sf.getString("layer"), 12)
+        .setAttrWithMinzoom("oneway", sf.getString("oneway"), 14)
+        // DEPRECATION WARNING: Marked for deprecation in v4 schema, do not use these for styling
+        //                      If an explicate value is needed it should bea kind, or included in kind_detail
+        .setAttr("highway", highway)
+        .setMinPixelSize(0)
+        .setPixelTolerance(0)
+        .setZoomRange(min_zoom, max_zoom);
+
+      // Core Tilezen schema properties
+      if (kind_detail != "") {
+        feat.setAttr("pmap:kind_detail", kind_detail);
+      } else {
+        feat.setAttr("pmap:kind_detail", highway);
+      }
+
+      // Core OSM tags for different kinds of places
+      if (service != "") {
+        feat.setAttr("service", service);
+      }
+
+      if (sf.hasTag("highway", "motorway_link", "trunk_link", "primary_link", "secondary_link",
         "tertiary_link")) {
         feat.setAttr("pmap:link", 1);
       }
 
-      if (sourceFeature.hasTag("bridge", "yes")) {
+      // Set "brunnel" (bridge / tunnel) property where "level" = 1 is a bridge, 0 is ground level, and -1 is a tunnel
+      // Because of MapLibre performance and draw order limitations, generally the boolean is sufficent
+      // See also: "layer" for more complicated ±6 layering for more sophisticated graphics libraries
+      if (sf.hasTag("bridge") && !sf.hasTag("bridge", "no")) {
         feat.setAttrWithMinzoom("pmap:level", 1, 12);
-      } else if (sourceFeature.hasTag("tunnel", "yes")) {
+      } else if (sf.hasTag("tunnel") && !sf.hasTag("tunnel", "no")) {
         feat.setAttrWithMinzoom("pmap:level", -1, 12);
       } else {
         feat.setAttrWithMinzoom("pmap:level", 0, 12);
       }
+
+      OsmNames.setOsmNames(feat, sf, min_zoom_names);
     }
   }
 
   @Override
   public List<VectorTile.Feature> postProcess(int zoom, List<VectorTile.Feature> items) {
+    if (zoom < 12) {
```

Comment:
Brandon fixed a bug near here, watch for merge conflicts / pull that in now...

Reply:
I've merged that PR.